### ISSUE TITLE
Use ECMAScript regexes in `builtins.{match,split}`

### ISSUE THREAD

**Issue by taktoa** [NixOS/nix#1520]
Allowing non-capturing groups in `builtins.{split,match}` would be very useful; currently it can be quite annoying to actually make use of the data output by a `builtins.match`, since you can't just eliminate all the irrelevant match groups. It would also be potentially good for performance.

Allowing some kind of unicode match would also be very useful.

Both of these would be satisfied by switching to `std::regex_constants::ECMAScript` from `std::regex_constants::extended`. This also has the advantage of being [part](http://en.cppreference.com/w/cpp/regex/ecmascript) of the C++ specification, instead of being specified in a separate standard (IEEE Std 1003.1-2008) that may or may not be what was actually implemented.

We need to determine if the existing uses of `builtins.match` are backwards-compatible with this regex format (though considering that we already changed regex libraries once I suspect it isn't a big issue).

**Comment by bb010g** [comment]
Is anything desired before moving forward with this, e.g. finding `split` or `match` usages in Nixpkgs or elsewhere that would change semantics?

**Comment by edolstra** [comment]
Indeed it would be good to know how much breakage this change would cause.

**Comment by taktoa** [comment]
Maybe we could ship a version of Nix with some kind of disable-able warning (i.e.: so that all users replace `builtins.match ...` with `silenceMatchMigration (builtins.match ...)`), and then ship another version that removes the old behavior and issues a warning if `silenceMatchMigration` is seen (so that there is a warning for people who added the `silence` without migrating to ECMAScript regexes).

[This GitHub search](https://github.com/search?l=Nix&p=1&q=%22builtins.match%22&type=Code) gives a good idea of all the uses of `builtins.match`

[This GitHub search](https://github.com/search?l=Nix&p=1&q=%22builtins.split%22&type=Code) gives a good idea of all the uses of `builtins.split`

**Comment by edolstra** [comment]
No, we can't do that. Users need to be able to depend on the semantics of built-in functions not suddenly changing. Maybe we could add an `builtins.ecmaMatch`, but I'm not convinced it's really useful.
